File: Parciales y simulacros/Python/Parcial 1C2024 Comision B/parcial.py
from queue import Queue as Cola

"""
Ejercicio 1) Juego del gallina (3 puntos)

El juego del gallina es una competición en la que dos participantes conducen un vehículo en dirección al del contrario; 
si alguno se desvía de la trayectoria de choque pierde y es humillado por comportarse como un "gallina". 
Se hizo un torneo para ver quién es el menos gallina. Juegan todos contra todos una vez y van sumando puntos, o restando. 
Si dos jugadores juegan y se chocan entre sí, entonces pierde cada uno 5 puntos por haberse dañado. 
Si ambos jugadores se desvían, pierde cada uno 10 puntos por gallinas. 
Si uno no se desvía y el otro sí, el gallina pierde 15 puntos por ser humillado y el ganador suma 10 puntos! 
En este torneo, cada persona que participa tiene una estrategia predefinida para competir: o siempre se desvía, o nunca lo hace. 
Se debe programar la función 'torneo_de_gallinas' que recibe un diccionario (donde las claves representan los nombres de los 
participantes que se anotaron en el torneo, y los valores sus respectivas estrategias) y devuelve un diccionario con los puntajes 
obtendidos por cada jugador.

problema torneo_de_gallinas (in estrategias: dict⟨String,String⟩) : dict⟨String,Z⟩ {
  requiere: {estrategias tiene por lo menos 2 elementos (jugadores)}
  requiere: {Las claves de estrategias tienen longitud mayor a 0}
  requiere: {Los valores de estrategias sólo pueden ser los strings "me desvio siempre" ó "me la banco y no me desvio"}
  asegura: {Las claves de res y las claves de estrategias son iguales}
  asegura: {para cada jugador p perteneciente a claves(estrategias), res[p] es igual a la cantidad de puntos que obtuvo al finalizar 
            el torneo, dado que jugó una vez contra cada otro jugador}
}
"""

def torneo_de_gallinas(estrategias: dict[str, str]) -> dict[str, int]:
    res: dict[str, int] = {}
    return res


"""
Ejercicio 2) Cola en el Banco (1 puntos)

En el banco ExactaBank los clientes hacen cola para ser atendidos por un representante. Los clientes son representados por las
tuplas (nombre, tipo afiliado) donde la primera componente es el nombre y el tipo afiliado puede ser "comun" o "vip".
Se nos pide implementar una función en python que dada una cola de clientes del banco, devuelva una nueva cola con los mismos 
clientes pero en donde los clientes vip estan primero que los clientes comunes manteniendo el orden original de los clientes vips 
y los comunes entre sí.

problema reordenar_cola_priorizando_vips (in filaClientes: Cola⟨String x String⟩) : Cola⟨String⟩ {
  requiere: {La longitud de los valores de la primera componente de las tuplas de la cola filaClientes es mayor a 0}
  requiere: {Los valores de la segunda componente de las tuplas de la cola filaClientes son "comun" o "vip" }
  requiere: {No hay dos tuplas en filaClientes que tengan la primera componente iguales entre sí }
  asegura: {todo valor de res aparece como primera componente de alguna tupla de filaClientes}
  asegura: {|res| = |filaCliente|}
  asegura: {res no tiene elementos repetidos}
  asegura: {No hay ningun cliente "comun" antes que un "vip" en res}
  asegura: {Para todo cliente c1 y cliente c2 de tipo "comun" pertenecientes a filaClientes si c1 aparece antes que c2 en filaClientes 
            entonces el nombre de c1 aparece antes que el nombre de c2 en res}
  asegura: {Para todo cliente c1 y cliente c2 de tipo "vip" pertenecientes a filaClientes si c1 aparece antes que c2 en filaClientes
            entonces el nombre de c1 aparece antes que el nombre de c2 en res}
}
"""
def copia_cola(cola: Cola[tuple[str, str]]) -> Cola[tuple[str, str]]:
  cola_auxiliar: Cola[tuple[str, str]] = Cola()
  cola_copiada: Cola[tuple[str, str]] = Cola()
  #Invertimos la cola pasada como parámetro
  while not cola.empty():
    elemento: tuple[str, str] = cola.get()
    cola_auxiliar.put(elemento)
  #Restauro la cola original
  while not cola_auxiliar.empty():
     otro_elemento: tuple[str, str] = cola_auxiliar.get()
     cola.put(otro_elemento)
     cola_copiada.put(otro_elemento)
  return cola_copiada

def reordenar_cola_priorizando_vips(filaClientes: Cola[tuple[str, str]]) -> Cola[str]:
  copia_filaClientes: Cola[tuple[str, str]] = copia_cola(filaClientes)
  afiliados_de_tipo_comun: Cola[str] = Cola()
  afiliados_de_tipo_vip: Cola[str] = Cola()
  res: Cola[str] = Cola()

  while not copia_filaClientes.empty():
    elemento: tuple[str, str] = copia_filaClientes.get()
    if elemento[1] == "comun":
      afiliados_de_tipo_comun.put(elemento[0])
    else:
      afiliados_de_tipo_vip.put(elemento[0])

  while not afiliados_de_tipo_vip.empty():
    cliente_vip: str = afiliados_de_tipo_vip.get()
    res.put(cliente_vip)
  while not afiliados_de_tipo_comun.empty():
    cliente_comun: str = afiliados_de_tipo_comun.get()
    res.put(cliente_comun)   
    
  return res

"""
Ejercicio 3) Sufijos que son palíndromos (2 puntos)

Decimos que una palabra es palíndromo si se lee igual de izquierda a derecha que de derecha a izquierda. 
Se nos pide programar en python la siguiente función:

problema cuantos_sufijos_son_palindromos(in texto:String) : Z {
  requiere: -
  asegura: {res es igual a la cantidad de palíndromos que hay en el conjunto de sufijos de texto}
}
Nota: un sufijo es una subsecuencia de texto que va desde una posición cualquiera hasta el final de la palabra. 
Ej: "Diego", el conjunto de sufijos es: "Diego", "iego","ego","go", "o". 
Para este ejercicio no consideraremos a "" como sufijo de ningun texto.
"""
def es_palindromo(texto: str) -> bool:
  for i in range(0, len(texto)):
    if i <= len(texto)-i-1:
      if texto[i] != texto[len(texto)-i-1]:
        return False
  return True

def cant_palindromos(subsecuencias: list[str]) -> int:
  cantidad: int = 0
  for i in range(len(subsecuencias)):
    if es_palindromo(subsecuencias[i]):
      cantidad+=1
  return cantidad

def cuantos_sufijos_son_palindromos(texto: str) -> int:
  sufijos: list[str] = []
  for i in range(len(texto)):
    sufijo: str = ""
    for j in range(i, len(texto)):
      sufijo = sufijo+texto[j]
    sufijos.append(sufijo)
  return cant_palindromos(sufijos)

"""
Ejercicio 4) Ta-Te-Ti-Facilito (2 puntos)

Ana y Beto juegan al Ta-Te-Ti-Facilito. El juego es en un tablero cuadrado de lado entre 5 y 10. Cada jugador va poniendo su ficha en cada turno.
Juegan intercaladamente y comienza Ana. Ana pone siempre una 'X' en su turno y Beto pone una 'O' en el suyo. 
Gana la persona que logra poner 3 fichas suyas consecutivas en forma vertical. Si el tablero está completo y no ganó nadie, 
entonces se declara un empate. El tablero comienza vacío, representado por ' ' en cada posición.
Notar que dado que juegan por turnos y comienza Ana poniendo una 'X' se cumple que la cantidad de 'X' es igual a la cantidad de 'O' o bien
la cantidad de 'X' son uno más que la cantidad de 'O'.
Se nos pide implementar una función en python 'problema quien_gano_el_tateti_facilito' que determine si ganó alguno, o si Beto hizo trampa 
(puso una 'O' cuando Ana ya había ganado).

problema quien_gano_el_tateti_facilito(in tablero:seq⟨seq⟨Char⟩) : Z {
  requiere: {tablero es una matriz cuadrada}
  requiere: {5<=|tablero[0]|<= 10}
  requiere: {tablero sólo tiene 'X', 'O' y ' ' (espacio vacío) como elementos}
  requiere: {En tablero la cantidad de 'X' es igual a la cantidad de 'O' o bien la cantidad de 'X' es uno más que la cantidad de 'O'}
  asegura: {res = 1 <==> hay tres 'X' consecutivas en forma vertical(misma columna) y no hay tres 'O' consecutivas en forma 
            vertical (misma columna) }
  asegura: {res = 2 <==> hay tres 'O' consecutivas en forma vertical (misma columna) y no hay tres 'X' consecutivas en forma 
            vertical (misma columna) }
  asegura: {res = 0 <==> no hay tres 'O' ni hay tres 'X' consecutivas en forma vertical}
  asegura: {res = 3 <==> hay tres 'X' y hay tres 'O' consecutivas en forma vertical (evidenciando que beto hizo trampa)}
}
"""
def quien_gano_el_tateti_facilito(tablero: list[list[str]]) -> int:
  for i in range(len(tablero[0])):
    contador_X: int = 0
    contador_O: int = 0
    for j in range(len(tablero)):
      if tablero[j][i] == "X":
        contador_X+=1
      elif tablero[j][i] == "O":
        contador_O+=1
      if contador_X == 3 and contador_O == 3:
        return 3
      elif contador_X == 3:
        return 1
      elif contador_O == 3:
        return 2
  return 0


"""
Ejercicio 5) Preguntas teóricas (2 puntos)
Conteste marcando la opción correcta.

A) ¿Qué es una variable en Python? (0.75 punto)

1) Una función que devuelve valores aleatorios.
2) Un contenedor para almacenar datos que puede cambiar durante la ejecución del programa.
3) Un tipo de dato que solo puede contener números enteros.

B) ¿Cuál es la función de un ciclo en Python? (0.75 punto)

1) Ejecutar un conjunto de instrucciones una sola vez.
2) Ejecutar repetidamente un conjunto de instrucciones hasta que una condición se evalúe como falsa.
3) Definir una constante que no puede ser cambiada.

C) ¿Cuál es la finalidad de un Control Flow Graph en testing? (0.5 punto)

1) Identificar todas las posibles salidas de un programa.
2) Visualizar todos los posibles caminos de ejecución para asegurar la cobertura completa del código.
3) Determinar los puntos de entrada y salida del programa.
"""

#Ejercicio 5 A) Un contenedor para almacenar datos que puede cambiar durante la ejecución del programa.
#Ejercicio 5 B) Ejecutar repetidamente un conjunto de instrucciones hasta que una condición se evalúe como falsa.
#Ejercicio 5 C) Visualizar todos los posibles caminos de ejecución para asegurar la cobertura completa del código.

File: Parciales y simulacros/Haskell/Parcial 1C2024 Comision B/Parcial.hs
module Parcial where

{-Enunciado
Una reconocida empresa de comercio electrónico nos pide desarrollar un sistema de stock de mercadería. 
El conjunto de mercaderías puede representarse con una secuencia de nombres de los productos donde puede haber productos repetidos. 
El stock puede representarse como una secuencia de tuplas de dos elementos, donde el primero es el nombre del producto y el segundo 
es la cantidad que hay en stock (en este caso no hay nombre de productos repetidos). 
También se cuenta con una lista de precios de productos representada como una secuencia de tuplas de dos elementos, donde el primero 
es el nombre del producto y el segundo es el precio.
Para implementar este sistema nos enviaron las siguientes especificaciones y nos pidieron que hagamos el desarrollo enteramente en Haskell, 
utilizando los tipos requeridos y solamente las funciones que se ven en la materia Introducción a la Programación / Algoritmos y Estructuras 
de Datos I (FCEyN-UBA).
-}

{-Ejercicio 1 (2 puntos)
problema generarStock (productos: seq⟨seq⟨Char⟩⟩): seq⟨seq⟨Char⟩ x Z⟩ {
    requiere: {True}
    asegura: {La longitud de res es igual a la cantidad de productos distintos que hay en productos}
    asegura: {Para cada producto que pertenece a productos existe un i tal que 0 <= i < |res| y res[i]_0 = producto y res[i]_1 es igual a la
              cantidad de veces que aparece producto en productos}
}
-}
longitud :: (Eq t) => [t] -> Int
longitud lista | lista == [] = 0
               | otherwise = 1 + longitud(tail lista)

cantApariciones :: [String] -> String -> Int
cantApariciones productos producto | longitud(productos) == 0 = 0
                                   | (head productos) == producto = 1 + cantApariciones (tail productos) producto
                                   | otherwise = cantApariciones (tail productos) producto

pertenece :: (Eq t) => [t] -> t -> Bool
pertenece lista elemento | longitud(lista) == 0 = False
                         | otherwise = (head lista) == elemento || pertenece (tail lista) elemento

generoStock :: [String] -> [String] -> [(String, Int)]
generoStock productos lista | longitud(productos) == 0 = []
                            | not(pertenece lista (head productos)) = ((head productos), cantApariciones productos (head productos)):generoStock (tail productos) ((head productos):lista)
                            | otherwise = generoStock (tail productos) lista

generarStock :: [String] -> [(String, Int)]
generarStock productos = generoStock productos []

{-Ejercicio 2 (2 puntos)
problema stockDeProducto (stock:seq⟨seq⟨Char⟩ x Z⟩, producto: seq⟨Char⟩) : Z {
    requiere: {No hay productos repetidos en stock}
    requiere: {Todas las cantidades de los productos que hay en stock son mayores a cero}
    asegura: {(res = 0 y producto no se encuentra en el stock) o (existe un i tal que 0 <= i < |stock| y producto = stock[i]_0 y
               res = stock[i]_1)}
}
-}
listaDeProductosStock :: [(String, Int)] -> [String]
listaDeProductosStock stock | longitud(stock) == 0 = []
                            | otherwise = fst(head stock):listaDeProductosStock (tail stock)

stockDeProducto :: [(String, Int)] -> String -> Int
stockDeProducto stock producto | not(pertenece (listaDeProductosStock(stock)) producto) = 0
                               | fst(head stock) == producto = snd(head stock)
                               | otherwise = stockDeProducto (tail stock) producto

{-Ejercicio 3 (2 puntos)
problema dineroEnStock (stock:seq⟨seq⟨Char⟩ x Z⟩, precios: seq⟨seq⟨Char⟩ x R⟩): R {
    requiere: {No hay productos repetidos en stock}
    requiere: {Todas las cantidades de los productos que hay en stock son mayores a cero}
    requiere: {No hay productos repetidos en precios}
    requiere: {Todos los precios de los productos son mayores a cero}
    requiere: {Todo producto en stock aparece en la lista de precios}
    asegura: {res es igual a la suma de los precios de todos los productos que están en stock multiplicado por 
              la cantidad de cada producto que hay en stock}
}
Para resolver este ejercicio pueden utilizar la función del Preludio de Haskell fromIntegral que dado un valor de tipo Int devuelve
su equivalente de tipo Float.
-}
listaDeProductosPrecios :: [(String, Float)] -> [String]
listaDeProductosPrecios precios | longitud(precios) == 0 = []
                                | otherwise = fst(head precios):listaDeProductosPrecios (tail precios)

precioDeProducto :: [(String, Float)] -> String -> Float
precioDeProducto precios producto | not(pertenece (listaDeProductosPrecios precios) producto) = 0
                                  | fst(head precios) == producto = snd(head precios)
                                  | otherwise = precioDeProducto (tail precios) producto

multiplicoCantConPrecio :: Int -> Float -> Float
multiplicoCantConPrecio cantidad precio = (fromIntegral cantidad)*precio

dineroEnStock :: [(String, Int)] -> [(String, Float)] -> Float
dineroEnStock stock precios | longitud(stock) == 0 = 0.0
                            | otherwise = (multiplicoCantConPrecio (stockDeProducto stock (fst(head stock))) (precioDeProducto precios (fst(head stock)))) + dineroEnStock (tail stock) (tail precios)

{-Ejercicio 4 (3 puntos)
problema aplicarOferta (stock:seq⟨seq⟨Char⟩ x Z⟩, precios: seq⟨seq⟨Char⟩ x R⟩: seq⟨seq⟨Char⟩ x R⟩) {
    requiere: {No hay productos repetidos en stock}
    requiere: {Todas las cantidades de los productos que hay en stock son mayores a cero}
    requiere: {No hay productos repetidos en precios}
    requiere: {Todos los precios de los productos son mayores a cero}
    requiere: {Todo producto en stock aparece en la lista de precios}
    asegura: {|res| = | precios|}
    asegura: {Para todo 0 <= i < |precios| si stockDeProducto(stock, precios[i]_0) > 10 entonces 
              res[i]_0 = precios[i]_0 y res[i]_1 = precios[i]_1 * 0,80}
    asegura: {Para todo 0 <= i < |precios| si stockDeProducto(stock, precios[i]_0) <= 10 entonces 
              res[i]_0 = precios[i]_0 y res[i]_1 = precios[i]_1}
}
-}
aplicarOferta :: [(String, Int)] -> [(String, Float)] -> [(String, Float)]
aplicarOferta stock precios | longitud(precios) == 0 = []
                            |(stockDeProducto (stock) (fst(head precios))) > 10 = ((fst(head precios)), (snd(head precios))*0.80):aplicarOferta (stock) (tail precios)
                            | otherwise = ((fst(head precios)), (snd(head precios))):aplicarOferta (stock) (tail precios)
{-Ejercicio 5 (1 punto). 
Conteste marcando la opción correcta.
En el contexto de la programación funcional, se llama polimorfismo:
1) Cuando una función puede invocarse con distintos tipos de datos sin tener que redefinirla.
2) Cuando una función puede invocarse con distintos tipos de datos teniendo que definirla para cada tipo de dato particular.
3) Cuando tengo un tipo de dato que puede ser usado para invocar a muchas funciones.
-}

--Ejercicio 5: Cuando una función puede invocarse con distintos tipos de datos sin tener que redefinirla.